src/django_components/util/template_tag.py

```diff
@@ -11,152 +16,36 @@
 from django_components.util.tag_parser import TagAttr, parse_tag
 
 
-@dataclass
-class TagSpec:
-    """Definition of args, kwargs, flags, etc, for a template tag."""
-
-    signature: inspect.Signature
-    """Input to the tag as a Python function signature."""
-    tag: str
-    """Tag name. E.g. `"slot"` means the tag is written like so `{% slot ... %}`"""
-    end_tag: Optional[str] = None
-    """
-    End tag.
-
-    E.g. `"endslot"` means anything between the start tag and `{% endslot %}`
-    is considered the slot's body.
-    """
-    flags: Optional[List[str]] = None
+# For details see https://github.com/EmilStenstrom/django-components/pull/902#discussion_r1913611633
+# and following comments
+def validate_params(
+    tag: str,
+    signature: inspect.Signature,
+    params: List["TagParam"],
+    extra_kwargs: Optional[Dict[str, Any]] = None,
+) -> None:
     """
-    List of allowed flags.
+    Validates a list of TagParam objects against this tag's function signature.
 
-    Flags are like kwargs, but without the value part. E.g. in `{% mytag only required %}`:
-    - `only` and `required` are treated as `only=True` and `required=True` if present
-    - and treated as `only=False` and `required=False` if omitted
-    """
-
-    def copy(self) -> "TagSpec":
-        sig_parameters_copy = [param.replace() for param in self.signature.parameters.values()]
-        signature = inspect.Signature(sig_parameters_copy)
-        flags = self.flags.copy() if self.flags else None
-        return self.__class__(
-            signature=signature,
-            tag=self.tag,
-            end_tag=self.end_tag,
-            flags=flags,
-        )
-
-    # For details see https://github.com/EmilStenstrom/django-components/pull/902
-    def validate_params(self, params: List["TagParam"]) -> Tuple[List[Any], Dict[str, Any]]:
```

Comment:
`TagSpec.validate_params` is now a standalone function https://github.com/EmilStenstrom/django-components/pull/910/files#diff-56d60a8b1d11d8ff86b912ddde6d8ee539f71a1fae917bc39a1fd049ea22705cR21